### PR TITLE
Make TraceWriter a Service object

### PR DESCRIPTION
**Pull request by ofrobots**

### CODE REVIEW

File: test/nocks.js [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+// In the future _=>true.
+function accept() {
+  return true;
+}
+
+function nockOAuth2(validator) {
+  validator = validator || accept;
+  return nock('https://accounts.google.com')
+      .post('/o/oauth2/token', validator)
+      .once()
```
Comment (matthewloring) [inline test/nocks.js]: Isn't `once` the default behavior on nocks?

File: test/nocks.js [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+}
+
+function nockProjectId(reply) {
+  return nock('http://metadata.google.internal')
+    .get('/computeMetadata/v1/project/project-id')
+    .once()
+    .reply(reply);
+}
+
+function nockPatchTraces(project, validator, reply, withError) {
```
Comment (matthewloring) [inline test/nocks.js]: Are all arguments after `project` optional here?

File: test/nocks.js [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    scope = scope.reply(reply || 200);
+  }
+  return scope;
+}
+
+module.exports = {
+  oauth2: nockOAuth2,
+  patchTraces: nockPatchTraces,
+  projectId: nockProjectId,
+};
```
Comment (matthewloring) [inline test/nocks.js]: '\n'

File: test/test-trace-writer.js [hunk 4]
```diff
@@ -18,94 +18,133 @@
...
-      samplingRate: 0,
-      forceNewAgent_: true
-    });
-    common.avoidTraceWriterAuth(agent);
-    cls.getNamespace().run(function() {
-      queueSpans(2, agent);
-      buf = common.getTraces(agent);
-      setTimeout(function() {
-        scope.done();
+function createFakeSpan(name) {
```
Comment (matthewloring) [inline test/test-trace-writer.js]: It would be nice for something like this to live in `common`.

File: test/nocks.js (hunk 1 above)
Comment (ofrobots) [inline test/nocks.js]: Being explicit is better though.

File: test/nocks.js (hunk 2 above)
Comment (ofrobots) [inline test/nocks.js]: Yes.

File: test/test-trace-writer.js (hunk 4 above)
Comment (ofrobots) [inline test/test-trace-writer.js]: Maybe. Depends on whether the other tests want the mock to be more functional. It is a tradeoff between what this set of unit tests needs vs reimplementing all the functionality of `SpanData`. `SpanData` has dependencies on `Agent` and `Logger` which need to be injected, perhaps it is better to inject these dependencies and use the real `SpanData` class?

File: src/trace-writer.js [hunk 5]
```diff
@@ -244,16 +249,19 @@ TraceWriter.prototype.flushBuffer_ = function(projectId) {
...
 
-  this.request_({
+  var options = {
     method: 'PATCH',
     uri: uri,
     body: json,
     headers: headers
-  }, function(err, response, body) {
+  };
+  that.request(options, function(err, body, response) {
```
Comment (kjin) [inline src/trace-writer.js]: Should this be `function(err, response, body)`? Both `this.request_` and `that.request` in the old and new revisions respectively are both assigned with a call to `util.makeAuthenticatedRequestFactory` (the latter in [`service.js`](https://github.com/GoogleCloudPlatform/google-cloud-node/blob/master/packages/common/src/service.js)).

File: test/nocks.js (hunk 3 above)
Comment (ofrobots) [inline test/nocks.js]: Done.

File: bin/run-test.sh [hunk 6]
```diff
@@ -39,7 +39,7 @@ function run {
 }
 
 # Run test/coverage
-for test in test/*.js test/plugins/*.js ;
```
Comment (matthewloring) [inline bin/run-test.sh]: We could also move the non-test config file into `test/fixtures`.

File: test/test-trace-writer.js (hunk 4 above)
Comment (matthewloring) [inline test/test-trace-writer.js]: This is something we can aim for long term. Nothing to change in this PR.

File: src/trace-writer.js (hunk 5 above)
Comment (ofrobots) [inline src/trace-writer.js]: I have been scratching my head about how this used to work before, but this does need to be `(err, body, response)`.

Comment (kjin) [inline src/trace-writer.js]: 🙃

**Pull request merged**